src/Microsoft.ML.LightGBM/LightGbmArguments.cs

```diff
@@ -24,12 +24,13 @@
 
 namespace Microsoft.ML.LightGBM
 {
-    public delegate void SignatureLightGBMBooster();
+    internal delegate void SignatureLightGBMBooster();
 
     [TlcModule.ComponentKind("BoosterParameterFunction")]
     public interface ISupportBoosterParameterFactory : IComponentFactory<IBoosterParameter>
     {
     }
+
     public interface IBoosterParameter
```

Comment:
>public [](start = 4, length = 6)

hide this one as well.

Reply:
I can do things like this right now:
`new Options.DartBooster.Arguments { }.CreateComponent(ML).UpdateParameters(new Dictionary<string,object>);`
does they make any sense for you?


---
In reply to: [255223941](https://github.com/dotnet/machinelearning/pull/2476#discussion_r255223941) [](ancestors = 255223941)

Reply:
The problem is `ISupportBoosterParameterFactory` is public. It is also used as a parameter of `Options` class which is public too. This does not allow us to make it internal. 

---
In reply to: [255258378](https://github.com/dotnet/machinelearning/pull/2476#discussion_r255258378) [](ancestors = 255258378,255223941)

Reply:
You can mark `public ISupportBoosterParameterFactory Booster = new TreeBooster.Arguments();` as internal and put in arguments attribute CmdLine only.
You can also create something like `public BoosterParameter BoosterParameters  = TreeBooster.Arguments()` and mark it EntryPoint only 
At least we can try to hide it.

---
In reply to: [255667107](https://github.com/dotnet/machinelearning/pull/2476#discussion_r255667107) [](ancestors = 255667107,255258378,255223941)

Reply:
`CreateComponent(ML)` and `UpdateParameters`. 

I am now using 
``` csharp 
public ISupportBoosterParameterFactory Booster = new TreeBooster.Arguments(); 
```
in the samples. So I think ISupportBoosterParameterFactory  can not be hidden.

---
In reply to: [256153739](https://github.com/dotnet/machinelearning/pull/2476#discussion_r256153739) [](ancestors = 256153739,255667107,255258378,255223941)

Reply:
I agree with Ivan I think you should be able to do as he suggested: make `mark public ISupportBoosterParameterFactory Booster` internal, and add a new public field that is of type `BoosterParameter`, or even `TreeBooster.Arguments`.

---
In reply to: [256250752](https://github.com/dotnet/machinelearning/pull/2476#discussion_r256250752) [](ancestors = 256250752,256153739,255667107,255258378,255223941)